[webkitpy] Respect webkitcorepy.Timeout in Executive.run_command

JonWBedard opened the pull request:
#### f3b42cb1ebac850ebe159e1a67e0446091d5d55e
<pre>
[webkitpy] Respect webkitcorepy.Timeout in Executive.run_command
<a href="https://bugs.webkit.org/show_bug.cgi?id=284648">https://bugs.webkit.org/show_bug.cgi?id=284648</a>
<a href="https://rdar.apple.com/141449707">rdar://141449707</a>

Reviewed by Sam Sneddon.

* Tools/Scripts/webkitpy/common/system/executive.py:
(Executive.run_command): Pass Timeout.difference() to &apos;wait()&apos; calls.

Canonical link: <a href="https://commits.webkit.org/287814@main">https://commits.webkit.org/287814@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3125ab4832783342f44c66b0be34645ab5d32ba7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80953 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/478 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34896 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/31939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8278 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/85482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/31939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84022 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/282 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73681 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/80410 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27838 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30397 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28387 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/86917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8183 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/86917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8360 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69513 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/86917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14779 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12555 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8144 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7981 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11501 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -391,17 +391,20 @@ def run_command(self,
+            exit_code = process.wait(timeout=Timeout.difference())
+            if exit_code is None:
+                self.kill_process(process.pid())
+                exit_code = 255
```
Is there a reason not to use `process.poll()` (or `process.wait()` and make sure it has been killed) here, and make sure we get the same exit code as the process?

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -391,17 +391,20 @@ def run_command(self,
             # wait() is not threadsafe and can throw OSError due to:
             # http://bugs.python.org/issue1731717
-            exit_code = process.wait()
+            exit_code = process.wait(timeout=Timeout.difference())
```
This should be a no-op, no?

`Popen.communicate` itself waits for the process to terminate or raises `TimeoutExpired`, so we're not gonna get a different result here.

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -391,17 +391,20 @@ def run_command(self,
         with process:
             if not string_to_communicate:
-                output = process.communicate()[0]
+                output = process.communicate(timeout=Timeout.difference())[0]
```
Do we not want to catch the `TimeoutException` here, so that we can handle that case (and, e.g., kill the process, rather than leaving it alive)?

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -391,17 +391,20 @@ def run_command(self,
+                output = process.communicate(timeout=Timeout.difference())[0]
             else:
-                output = process.communicate(string_utils.encode(string_to_communicate, encoding='utf-8'))[0]
+                output = process.communicate(string_utils.encode(string_to_communicate, encoding='utf-8'), timeout=Timeout.difference())[0]
```
Ditto.

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -391,17 +391,20 @@ def run_command(self,
             # http://bugs.python.org/issue1731717
-            exit_code = process.wait()
+            exit_code = process.wait(timeout=Timeout.difference())
+            if exit_code is None:
```
Pretty sure `exit_code` can never be `None`.

https://github.com/python/typeshed/blob/22e0699ac97a849c669e96685feb41ec6f34b55b/stdlib/subprocess.pyi#L2557 would tend to confirm this.

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -389,37 +389,42 @@ def run_command(self,
-                                           cwd=cwd)
+        finally:
+            if process.poll() is None:
+                process.kill()
```
Do we want to wait for it to actually be killed? (IIRC, on NT systems, killing a process can take a small amount of time.)

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -389,37 +389,42 @@ def run_command(self,
-                return exit_code
+        except subprocess.TimeoutExpired:
+            _log.debug('"%s" timed out after %.2fs' % (self.command_for_printing(args), time.time() - start_time))
+            exit_code = 255
```
Setting this to 255 seems… maybe surprising? I'd maybe just expect the exit code to be that of `SIGKILL`, which suggests not setting it here, but just setting it when `process.poll()` returns not-`None`.

JonWBedard committed:
[webkitpy] Respect webkitcorepy.Timeout in Executive.run_command
https://bugs.webkit.org/show_bug.cgi?id=284648
rdar://141449707

Reviewed by Sam Sneddon.

* Tools/Scripts/webkitpy/common/system/executive.py:
(Executive.run_command): Pass Timeout.difference() to 'wait()' calls.

Canonical link: https://commits.webkit.org/287814@main
